pkg/providers/instance/suite_test.go

```diff
@@ -291,4 +293,90 @@ var _ = Describe("InstanceProvider", func() {
 		Expect(len(interfaces)).To(Equal(1))
 		Expect(interfaces[0].Name).To(Equal(managedNic.Name))
 	})
+	Context("CSExtension", func() {
+		var originalOptions *options.Options
+
+		BeforeEach(func() {
+			originalOptions = options.FromContext(ctx)
+			ctx = options.ToContext(
+				ctx,
+				test.Options(test.OptionsFields{
+					ProvisionMode: lo.ToPtr(consts.ProvisionModeBootstrappingClient),
+				}))
+		})
+
+		AfterEach(func() {
+			ctx = options.ToContext(ctx, originalOptions)
+		})
+
+		It("should create CSExtension with correct configuration", func() {
+			ExpectApplied(ctx, env.Client, nodePool, nodeClass)
+			pod := coretest.UnschedulablePod(coretest.PodOptions{})
+			ExpectProvisioned(ctx, env.Client, cluster, cloudProvider, coreProvisioner, pod)
+			ExpectScheduled(ctx, env.Client, pod)
+
+			Expect(azureEnv.VirtualMachinesAPI.VirtualMachineCreateOrUpdateBehavior.CalledWithInput.Len()).To(Equal(1))
+
+			Expect(azureEnv.VirtualMachineExtensionsAPI.VirtualMachineExtensionsCreateOrUpdateBehavior.CalledWithInput.Len()).To(Equal(2)) // One for CSE, one for AKS identifying extension
+
+			var cseExtension *armcompute.VirtualMachineExtension
+			for i := 0; i < 2; i++ {
+				input := azureEnv.VirtualMachineExtensionsAPI.VirtualMachineExtensionsCreateOrUpdateBehavior.CalledWithInput.Pop()
+				if *input.VirtualMachineExtension.Name == "cse-agent-karpenter" {
+					cseExtension = &input.VirtualMachineExtension
+					break
+				}
+			}
+			Expect(cseExtension).ToNot(BeNil())
+
+			Expect(*cseExtension.Properties.Type).To(Equal("CustomScript"))
+			Expect(*cseExtension.Properties.Publisher).To(Equal("Microsoft.Azure.Extensions"))
+			Expect(*cseExtension.Properties.TypeHandlerVersion).To(Equal("2.0"))
+			Expect(cseExtension.Properties.ProtectedSettings).ToNot(BeNil())
+
+			protectedSettings, ok := cseExtension.Properties.ProtectedSettings.(map[string]interface{})
+			Expect(ok).To(BeTrue())
+			Expect(protectedSettings["commandToExecute"]).ToNot(BeEmpty())
+
+			Expect(cseExtension.Tags).ToNot(BeNil())
+			Expect(cseExtension.Tags[instance.NodePoolTagKey]).ToNot(BeNil())
+			Expect(*cseExtension.Tags[instance.NodePoolTagKey]).To(Equal(nodePool.Name))
+		})
+
+		It("should create Windows CSExtension with correct configuration", func() {
+			nodeClass.Spec.ImageFamily = lo.ToPtr("Windows2019")
+			ExpectApplied(ctx, env.Client, nodePool, nodeClass)
+
+			pod := coretest.UnschedulablePod(coretest.PodOptions{})
+			ExpectProvisioned(ctx, env.Client, cluster, cloudProvider, coreProvisioner, pod)
+			ExpectScheduled(ctx, env.Client, pod)
+
+			Expect(azureEnv.VirtualMachinesAPI.VirtualMachineCreateOrUpdateBehavior.CalledWithInput.Len()).To(Equal(1))
+
```

Comment:
We don't support windows, this was some feedback I had on the 2a code that it may make things confusing to put all the wiring. 

 [FAILED] Expected success, but got an error:
      <*errors.StatusError | 0xc000dfe140>: 
      AKSNodeClass.karpenter.azure.com "roarerclever-51-7wy0bfzllv" is invalid: [spec.imageFamily: Unsupported value: "Windows2019": supported values: "Ubuntu2204", "AzureLinux", <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation
      
      Wont work with windows.

Reply:
Deleted the Windows related test.